### PR TITLE
[MRG+1] Don't check `iterable()` before `len()`.

### PR DESCRIPTION
**Pull request by anntzer**
... because the former does not imply the latter anyways, e.g.
generators are iterables but unsized.

Now `plot(zip([1, 2], [3, 4]))` (py3) raises `RuntimeError: matplotlib
does not support generators` (from `cbook.safe_first_element`) which is
probably the intended exception, rather than `TypeError: object of type
'zip' has no len()`.  Perhaps this exception should be changed to a
`TypeError`, by the way...

### CODE REVIEW

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -409,42 +409,27 @@ def _set_loc(self, loc):
...
-
-        self._loc_real = loc
         self.stale = True
 
     def _get_loc(self):
         return self._loc_real
 
     _loc = property(_get_loc, _set_loc)
 
-    def _findoffset_best(self, width, height, xdescent, ydescent, renderer):
```
Comment (phobson) [inline lib/matplotlib/legend.py]: are these changes related to the title of the PR?

Comment (anntzer) [inline lib/matplotlib/legend.py]: The idea was to fix the line a bit below (`if iterable(self._loc) and len(self._loc) == 2`) but as you can see the new logic is just to see whether `_loc` is a 0-9 location and otherwise unpack it, so it seemed reasonable to squash the `if ... elif ... else` into a single function rather than keeping them artificially separated.

File: lib/matplotlib/legend.py [hunk 2]
```diff
@@ -409,42 +409,27 @@ def _set_loc(self, loc):
...
-            x, y = bbox.x0 + bbox.width * fx, bbox.y0 + bbox.height * fy
-        else:
+        if self._loc == 0:  # "best".
+            x, y = self._find_best_position(width, height, renderer)
+        elif self._loc in Legend.codes.values():  # Fixed location.
             bbox = Bbox.from_bounds(0, 0, width, height)
             x, y = self._get_anchored_bbox(self._loc, bbox,
                                            self.get_bbox_to_anchor(),
                                            renderer)
+        else:  # Axes or figure coordinates.
```
Comment (story645) [inline lib/matplotlib/legend.py]: is the code set up such that if self._loc is neither 0 nor in Legend.codes.values(), then it has to be iterable of len 2?

File: lib/matplotlib/tests/test_collections.py [hunk 3]
```diff
@@ -646,12 +646,12 @@ def test_lslw_bcast():
     col.set_linestyles(['-', '-'])
     col.set_linewidths([1, 2, 3])
 
-    assert col.get_linestyles() == [(None, None)] * 6
-    assert col.get_linewidths() == [1, 2, 3] * 2
+    assert_equal(col.get_linestyles(), [(None, None)] * 6)
```
Comment (story645) [inline lib/matplotlib/tests/test_collections.py]: any reason you're using numpy testing over py.test asserts here?

File: lib/matplotlib/units.py [hunk 4]
```diff
@@ -157,11 +157,15 @@ def get_converter(self, x):
...
-                return converter
+        if converter is None:
+            try:
+                thisx = safe_first_element(x)
+            except (TypeError, StopIteration):
+                pass
+            else:
+                if classx and classx != getattr(thisx, '__class__', None):
+                    converter = self.get_converter(thisx)
+                    return converter
```
Comment (story645) [inline lib/matplotlib/units.py]: the converter is always returned at the bottom of this function (it's literally the next line) so is it worth putting here? And a general question/qualm on the multiple returns: while I know PEP8 is lenient on them, is there a good reason for them here where everything is bundled in if statements anyway?

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/legend.py]: The condition effectively is the same as before.

Previous impl:
- if loc == 0, use findoffset_best, which directly calls find_best_position.
- otherwise, use findoffset_loc:
  - if len(loc) == 2, use fixed position
  - otherwise, use get_anchored_bbox (fixed location, assumed to be in Legend.codes.values)

New impl:
- if loc == 0, use find_best_position
- if fixed location, in Legend.codes.values, use get_anchored_bbox
- otherwise, assume fixed position

It is correct that the error on invalid locations would be different though.  The previous implementation would raise an AssertionError (`assert loc in range(1, 11)  # called only internally` in get_anchored_bbox), the current one raises a TypeError (when trying to iterate over self._loc).

File: lib/matplotlib/units.py (hunk 4 above)
Comment (anntzer) [inline lib/matplotlib/units.py]: I tried to keep the logical flow of the previous implementation as much as possible -- I think the whole thing probably needs some refactor anyways, so there isn't much in favor of a piecemeal approach.

File: lib/matplotlib/tests/test_collections.py (hunk 3 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_collections.py]: because `col.get_linewidths` can now return an array rather than a list, due to the switch from
`self._us_lw = self._get_value(lw)`
to
`self._us_lw = np.atleast_1d(np.asarray(lw))`

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (story645) [inline lib/matplotlib/legend.py]: :/ Should it be raising a `ValueError` since that's what's actually going wrong? I think I get now why the old code checked for len first...

File: lib/matplotlib/units.py (hunk 4 above)
Comment (story645) [inline lib/matplotlib/units.py]: I get...refactoring units (and adding ScalerMappable support) is on my wishlist...

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/legend.py]: The previous version would raise AssertionError both for invalid single values (loc=99) and for invalid length values (loc=(1, 2, 3)).  The new implementation now raises TypeError for the former and ValueError for the latter.  I think the new version is better, given that most functions can indeed raise both TypeError and ValueError when given an incorrect input, so you need to catch both anyways (whereas catching AssertionError is pretty rare).

File: lib/matplotlib/tests/test_collections.py (hunk 3 above)
Comment (NelleV) [inline lib/matplotlib/tests/test_collections.py]: Does `assert_equal` work on arrays? don't you have to use `assert_array_equal`?

Comment (anntzer) [inline lib/matplotlib/tests/test_collections.py]: It works.  See also https://github.com/numpy/numpy/issues/8457.
```
In [4]: np.testing.assert_equal(np.array([1, 2]), [1, 2])

In [5]: np.testing.assert_equal(np.array([1, 2]), [1, 3])
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-f3a660c3ad1f> in <module>()
----> 1 np.testing.assert_equal(np.array([1, 2]), [1, 3])

/usr/lib/python3.6/site-packages/numpy/testing/utils.py in assert_equal(actual, desired, err_msg, verbose)
    320     from numpy.lib import iscomplexobj, real, imag
    321     if isinstance(actual, ndarray) or isinstance(desired, ndarray):
--> 322         return assert_array_equal(actual, desired, err_msg, verbose)
    323     msg = build_err_msg([actual, desired], err_msg, verbose=verbose)
    324 

/usr/lib/python3.6/site-packages/numpy/testing/utils.py in assert_array_equal(x, y, err_msg, verbose)
    811     """
    812     assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
--> 813                          verbose=verbose, header='Arrays are not equal')
    814 
    815 def assert_array_almost_equal(x, y, decimal=6, err_msg='', verbose=True):

/usr/lib/python3.6/site-packages/numpy/testing/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision)
    737                                 names=('x', 'y'), precision=precision)
    738             if not cond:
--> 739                 raise AssertionError(msg)
    740     except ValueError:
    741         import traceback

AssertionError: 
Arrays are not equal

(mismatch 50.0%)
 x: array([1, 2])
 y: array([1, 3])
```
